webapi/src/main/scala/org/knora/webapi/responders/v2/resources/CheckObjectClassConstraints.scala

```diff
@@ -36,72 +39,82 @@ object CheckObjectClassConstraints {
     entityInfo: EntityInfoGetResponseV2,
     clientResourceIDs: Map[IRI, String],
     resourceIdForErrorMsg: IRI,
+    ontologyRepo: OntologyRepo,
   )(implicit
     stringFormatter: StringFormatter,
   ): Task[Unit] =
     for {
-      propertyInfo <- ZIO.succeed(entityInfo.propertyInfoMap(propertyIri))
-
-      // Don't accept link properties.
-      _ <- ZIO.when(propertyInfo.isLinkProp)(ZIO.fail(invalidPropUseLinkProp(resourceIdForErrorMsg, propertyIri)))
+      propertyInfo      <- ZIO.succeed(entityInfo.propertyInfoMap(propertyIri))
+      _                 <- ZIO.when(propertyInfo.isLinkProp)(ZIO.fail(invalidPropUseLinkProp(resourceIdForErrorMsg, propertyIri)))
+      objectConstraints <- propertysObjectConstraint(propertyIri, entityInfo, propertyInfo, ontologyRepo)
 
-      // Get the property's object class constraint. If this is a link value property, we want the object
-      // class constraint of the corresponding link property instead.
-      propertyInfoForObjectClassConstraint = if (propertyInfo.isLinkValueProp)
-                                               entityInfo.propertyInfoMap(propertyIri.fromLinkValuePropToLinkProp)
-                                             else
-                                               propertyInfo
+      // Check each value.
+      _ <-
+        ZIO.foreachDiscard(valuesToCreate.map(_.valueContent)) {
+          case linkValueContentV2: LinkValueContentV2 =>
+            // It's a link value.
+            for {
+              _ <-
+                ZIO.when(!propertyInfo.isLinkValueProp) {
+                  propertyRequiresValue(resourceIdForErrorMsg, propertyIri, objectConstraints)
+                }
+
+              // Does the resource that's the target of the link belongs to a subclass of the
+              // link property's object class constraint?
+              linkTargetClass     = linkTargetClasses(linkValueContentV2.referredResourceIri)
+              linkTargetClassInfo = entityInfo.classInfoMap(linkTargetClass)
+              _ <-
+                ZIO.when(!linkTargetClassInfo.allBaseClasses.exists(objectConstraints.contains)) {
+                  // No. If the target resource already exists, use its IRI in the error message.
+                  // Otherwise, use the client's ID for the resource.
+                  ZIO.fail(
+                    propertyInvalid(
+                      resourceIdForErrorMsg,
+                      linkValueContentV2,
+                      clientResourceIDs,
+                      propertyIri,
+                      objectConstraints,
+                    ),
+                  )
+                }
+            } yield ()
+
+          case other: ValueContentV2 =>
+            // It's not a link value. Check that its type is equal to the property's object class constraint.
+            ZIO.when(!objectConstraints.contains(other.valueType)) {
+              propertyRequiresValue(resourceIdForErrorMsg, propertyIri, objectConstraints)
+            }
```

Comment:
**suggestion (non-blocking):** how about extracting the value handling into a separate function?

Reply:
I am not a big fan either when it comes to having a for comprehension inside a case inside something else.

However, extracting the LinkValue would result in a function with nine params, which I am not a big fan of either and I think simply extracting the function would not have much benefit.

I leave this refactoring exercise to a future us.